### ISSUE TITLE
Weird colors with esp32-8048S070C

### ISSUE THREAD

**Issue by Andre-Schuiki** [rzeldent/esp32-smartdisplay#60]
Hi all,

i tried the demo application but can't get the screen to work correct.
My colors looking weird...any ideas what i'm doing wrong?

![image](https://github.com/rzeldent/esp32-smartdisplay/assets/4403123/a2a3713a-c6e4-4ea3-8de6-38abda5d4651)

This is my platformio.ini
```ini
; PlatformIO Project Configuration File
;
;   Build options: build flags, source filter
;   Upload options: custom upload port, speed and extra flags
;   Library options: dependencies, extra library storages
;   Advanced options: extra scripting
;
; Please visit documentation for the other options and examples
; https://docs.platformio.org/page/projectconf.html

[platformio]
default_envs = esp32-8048S070C

[env]
platform = espressif32
framework = arduino

monitor_speed = 115200
monitor_rts = 0
monitor_dtr = 0
monitor_filters = esp32_exception_decoder

# Partition scheme for OTA
board_build.partitions = min_spiffs.csv

build_flags =
    -Ofast
    -Wall
    -DCORE_DEBUG_LEVEL=ARDUHAL_LOG_LEVEL_VERBOSE
    # LVGL settings
    -DLV_CONF_PATH=${PROJECT_INCLUDE_DIR}/lv_conf.h

#lib_deps = rzeldent/esp32_smartdisplay
lib_deps =
    https://github.com/rzeldent/esp32-smartdisplay.git
    https://github.com/schreibfaul1/ESP32-audioI2S

[env:esp32-8048S070C]
board = esp32-8048S070C
```

**Comment by rzeldent** [comment]
Hi Andre,

Looks like the LV_COLOR16_SWAP is incorrect.

**Comment by Andre-Schuiki** [comment]
Yeah that's what i thought too but i already tried with swap enabled and get the same result.

I will try it again tomorrow and give you feedback.

Thanks for your help!

**Comment by rzeldent** [comment]
It can be that you need to rebuild the library. The define is only taken into account when compiling thew library (that is cached).
Retry it but delete the .pio directory first. It is also possible that you need to change the settings in Squareline to be able to compile...
This should have been solved in the 2.0.x versions?!

**Comment by Andre-Schuiki** [comment]
> It can be that you need to rebuild the library. The define is only taken into account when compiling thew library (that is cached). Retry it but delete the .pio directory first. It is also possible that you need to change the settings in Squareline to be able to compile... This should have been solved in the 2.0.x versions?!

Ok i have tried with LV_COLOR_16_SWAP 1 but i get the same result. (i also have deleted the .pio folder before compiling)

**Comment by IanSC** [comment]
I have the same issue as this. Edited lv_conf.h and changed

`#define LV_COLOR_16_SWAP 0`

Did a PIO: Clean by clicking the trash can before uploading. Got the correct photo color,
but text are garbled and the colors for widgets are off.
eg. green border, greenish background, avocado button

### esp32-8048S070C

<img src="https://github.com/rzeldent/esp32-smartdisplay/assets/8453489/13ed070f-42b6-42e2-9d36-65bd6f9f33f5" width="100%">

<img src="https://github.com/rzeldent/esp32-smartdisplay/assets/8453489/78eaab25-cb50-4731-9bb1-7932a8c83c17" width="100%">

<img src="https://github.com/rzeldent/esp32-smartdisplay/assets/8453489/3797e2a2-7509-45ac-bf6a-ced6b84430f7" width="100%">
   
<br/><br/>
Tried it on another device and it appears correct (set the swap var to 1 since SPI).

### esp32-3248S035C

<img src="https://github.com/rzeldent/esp32-smartdisplay/assets/8453489/44406c2d-46dd-4d28-97e2-c8e1199591ce" width="50%">

<br/><br/>
Not sure if related, but the downloaded library does not contain ST7701 files.

Thanks for the help and the great library!

**Comment by rzeldent** [comment]
Hi IanSC,

Maybe the 16 bits color array is swapped, that's what I suspect. You can try to swap it by manually by setting in lvgl_rgb_panel.c the color 16 swap, just for testing:
```
// Create direct_io panel handle
#if LV_COLOR_16_SWAP != 0
```

Ordered the 8048S070C and will hopefully be able to reproduce it and correct it.

**Comment by IanSC** [comment]
Hi rzeldent,

**A**: `lv_conf.h`   `#define LV_COLOR_16_SWAP <value>`
**B**: `lvgl_rgb_panel.c, line 28`   `#if LV_COLOR_16_SWAP <operator> 0`

|  A < value > | B < operator > | Image | Text | Colors |
|---|---|---|---|---|
| 1 | != | inverted | garbled | off |
| 0 | != | okay | garbled | off |
| 1 | == | inverted | garbled | off |
| 0 | == | okay | garbled | off |

Thanks for checking this out.

**Comment by rzeldent** [comment]
Hi LanC,

Thanks for testing this. I was hoping it was just a typo... So basically the text is always garbled.... 
Have you tried inverting the image? esp_lcd_panel_invert_color(handle, true);

**Comment by IanSC** [comment]
Hi rzeldent,

Took me a while to figure out where to the add the line of code.

I also tried the other cases from your previous suggestion.
I doubted there would be any difference since it's a text issue not color,
but surprisingly it got much better.

Thanks again!

in `lvgl_rgb_panel.c`, added:
`ESP_ERROR_CHECK(esp_lcd_panel_invert_color(panel_handle, true));`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  at the end of the function `lvgl_lcd_init()`

**A**: `lv_conf.h`   `#define LV_COLOR_16_SWAP <value>`
**B**: `lvgl_rgb_panel.c, line 28`   `#if LV_COLOR_16_SWAP <operator> 0`

| CASE |  A < value > | B < operator > | Image | Text | Colors | Shade | Note |
|:---|:---|:---|:---|:---|:---|:---|:---|
| 1 | 1 | == | inverted | garbled | off | none | |
| 2 | 0 | == | wrong | garbled | off | purple | |
| 3 | 1 | != | wrong | okay | off | dark purple | |
| 4 | 0 | != | inverted | okay | okay | none | FPS overlay seems off |

### CASE 1
<img src="https://github.com/rzeldent/esp32-smartdisplay/assets/8453489/eae5beba-ea37-45ef-84ea-2e5cc59039c4" width="100%">

### CASE 2
<img src="https://github.com/rzeldent/esp32-smartdisplay/assets/8453489/cd134a70-003e-40f1-969d-798bf8508077" width="100%">

### CASE 3
<img src="https://github.com/rzeldent/esp32-smartdisplay/assets/8453489/525e2ac7-eba5-4f0b-a59d-f95796252bf5" width="100%">

### CASE 4
<img src="https://github.com/rzeldent/esp32-smartdisplay/assets/8453489/e15f01cb-86c8-47da-ac09-91fd10bac657" width="100%">

**Comment by IanSC** [comment]
Additional Test:
Just reporting here to help solve any potential issues.
I don't remember the colors being so far off with the built in app
that came with the device.

There are not bug reports, just observations.
Big screen is yellowish and less brighter.
Added:
```
void setup() {
    ...
    smartdisplay_init();
    // smartdisplay_lcd_set_backlight(1.0f); // screen flickers
    // smartdisplay_lcd_set_backlight(0.9f); // screen flickers
    smartdisplay_lcd_set_backlight(0.5f);    // screen flickers
    ...
    tutorialCode();
}
void loop() {
    lv_timer_handler();
}
```
Changing the brightness causes the screen to flicker and become unstable, even if set to 0.5f
which is the same inside smartdisplay_init(). I might be calling it in the wrong place.

Button test seems normal, except for overlay.
Solid RGB colors are off.
<br/>

![IMG_20240108_213810](https://github.com/rzeldent/esp32-smartdisplay/assets/8453489/783b98c5-a22d-460a-ab82-1cfdc1175ba0)
![IMG_20240108_222219](https://github.com/rzeldent/esp32-smartdisplay/assets/8453489/8dd51efb-8771-423f-b0ca-0f18e26c59c9)
![IMG_20240108_222239](https://github.com/rzeldent/esp32-smartdisplay/assets/8453489/ef261960-27a4-46d7-8d13-2abd4293461d)
![IMG_20240108_222254](https://github.com/rzeldent/esp32-smartdisplay/assets/8453489/20075e95-988d-4c7f-8c97-b11379033fc7)

**Comment by IanSC** [comment]
Hi rzeldent,

Based on:
Esp32-8048S070 display, different version
https://github.com/rzeldent/esp32-smartdisplay/issues/90

I tried the settings for v1.3 and the text and colors are correct.

For noobs like me:
- make a copy of ``esp32-8048S070C.json`` inside ``board`` folder
- rename as, say ``esp32-8048S070C-v1.3.json``
- edit and apply suggested changes in Issue #90
- edit ``platformio.ini`` and use ``board = esp32-8048S070C-v1.3``
- rebuild

**Comment by IanSC** [comment]
Tried LVGL examples with the changes and it works.

Thanks to @rzeldent !

So guys, let's support or buy this guy a beer, without his efforts these boards will be 
useless or extremely difficult to use. Looking at you Sunton.

**Comment by rzeldent** [comment]
Thanks!

BTW in the develop the v1.3 board will be the default. Think the v1.1 was some typo.

**Comment by MrOptifine** [comment]
Any idea on how to fix the screen flickers?

**Comment by rzeldent** [comment]
Hi Optifine,

I think the PCLK needs tweaking....
